File: taller/AMarotoChivite-taller/AMarotoChivite-taller-Java/src/main/java/simulacionTaller/factories/WorkShopFactory.java
package simulacionTaller.factories;

import simulacionTaller.models.PersonBrench;

import static simulacionTaller.utils.randomInfo.setDependentsOfBosses;

public class WorkShopFactory {

    /**
     * Creará un taller con los porcentajes dados de cada miembro, el resto serán trabajadores normales,
     * y una pequeña probabilidad de aparecer un MultiTrabajador (chapista y electricista)
     *
     * @return taller con todos los miembros
     */
    public static PersonBrench[] create(int quantityPerson, int bossPercent, int electricianPercent,
                                        int platerPercent) {
        PersonBrench[] workShop = new PersonBrench[quantityPerson];

        int quantityBoss = ((quantityPerson * bossPercent) / 100);
        int quantityElectrician = ((quantityPerson * electricianPercent) / 100);
        int quantityPlater = ((quantityPerson * platerPercent) / 100);

        int noWorkerAssign = quantityPerson
                - quantityBoss
                - quantityElectrician
                - quantityPlater;

        int quantityMulti = getQuantityMulti(noWorkerAssign);

        for (int i = 0; i < workShop.length; i++) {
            if (i < quantityBoss) {
                workShop[i] = BossFactory.create();
            } else if (i < quantityBoss + quantityElectrician) {
                workShop[i] = ElectricianFactory.create();
            } else if (i < quantityBoss + quantityElectrician + quantityPlater) {
                workShop[i] = PlaterFactory.create();
            } else if (i < quantityBoss + quantityElectrician + quantityPlater + quantityMulti) {
                workShop[i] = MultiFactory.create();
            } else {
                workShop[i] = WorkerFactory.create();
            }
        }
        // Establecemos a cada jefe sus trabajadores
        setDependentsOfBosses(workShop);
        return workShop;
    }

    public static int getQuantityMulti(int restNoWorkerAssign) {
        int quantityMulti = 0;

        for (int i = 1; i <= restNoWorkerAssign; i++) {
            // Probabilidad de que salga un multi será de 10%
            int random = (int) (Math.random() * 10 + 1);
            if (random == 1) {
                quantityMulti++;
            }
        }

        return quantityMulti;
    }
}

File: taller/AMarotoChivite-taller/AMarotoChivite-taller-kotlin/src/main/kotlin/simulacionTaller/factories/WorkShopFactory.kt
package simulacionTaller.factories

import simulacionTaller.models.PersonBrench
import simulacionTaller.utils.setDependentsOfBosses

object WorkShopFactory {
    /**
     * Creará un taller con los porcentajes dados de cada miembro, el resto serán trabajadores normales,
     * y una pequeña probabilidad de aparecer un MultiTrabajador (chapista y electricista)
     *
     * @return taller con todos los miembros
     */
    @JvmStatic
    fun create(
        quantityPerson: Int, bossPercent: Int, electricianPercent: Int,
        platerPercent: Int
    ): Array<PersonBrench?> {
        val workShop = arrayOfNulls<PersonBrench>(quantityPerson)
        val quantityBoss = quantityPerson * bossPercent / 100
        val quantityElectrician = quantityPerson * electricianPercent / 100
        val quantityPlater = quantityPerson * platerPercent / 100
        val noWorkerAssign = (quantityPerson
                - quantityBoss
                - quantityElectrician
                - quantityPlater)
        val quantityMulti = getQuantityMulti(noWorkerAssign)
        for (i in workShop.indices) {
            if (i < quantityBoss) {
                workShop[i] = BossFactory.create()
            } else if (i < quantityBoss + quantityElectrician) {
                workShop[i] = ElectricianFactory.create()
            } else if (i < quantityBoss + quantityElectrician + quantityPlater) {
                workShop[i] = PlaterFactory.create()
            } else if (i < quantityBoss + quantityElectrician + quantityPlater + quantityMulti) {
                workShop[i] = MultiFactory.create()
            } else {
                workShop[i] = WorkerFactory.create()
            }
        }
        // Establecemos a cada jefe sus trabajadores
        setDependentsOfBosses(workShop)
        return workShop
    }

    @JvmStatic
    fun getQuantityMulti(restNoWorkerAssign: Int): Int {
        var quantityMulti = 0
        for (i in 1..restNoWorkerAssign) {
            // Probabilidad de que salga un multi será de 10%
            val random = (Math.random() * 10 + 1).toInt()
            if (random == 1) {
                quantityMulti++
            }
        }
        return quantityMulti
    }
}